[RFC] Bluetooth: Allow printing out HCI error codes as strings

rugeGerritsen opened the pull request:
When developing Bluetooth applications, you typically run into some errors. If you are an experienced Bluetooth developer, you would typically have an HCI error lookup table in your memory. Others might not.

This commit conditionally prints out error codes.

**Needs discussion**: 
* How to use them
* Where to use them
* Implementation details

This draft PR is created as a starting point for the discussion

rugeGerritsen commented on samples/bluetooth/peripheral_hr/src/main.c:
```diff
@@ -56,7 +56,7 @@ static void connected(struct bt_conn *conn, uint8_t err)
 static void disconnected(struct bt_conn *conn, uint8_t reason)
 {
-	printk("Disconnected (reason 0x%02x)\n", reason);
+	printk("Disconnected (reason " HCI_ERR_FMT ")\n", HCI_ERR(reason));
```
IMO, this looks a bit ugly, so I'm open for other suggestions!

alwa-nordic replied:
I have been trough a similar though process and decided to just print both. In the case of `CONFIG_BT_HCI_ERR_AS_STRING`, the function can be macroed out to the empty string, which I hope optimizes nicely.

https://github.com/zephyrproject-rtos/zephyr/blob/786b9a0ad405c3475382545aa29e3b815129e68a/subsys/bluetooth/shell/gatt.c#L1013-L1014

rugeGerritsen replied:
Good suggestion. I like that better than mine. It also makes it more clear that it that it is possible to enable and disable string representations

rugeGerritsen replied:
I would have hoped the compiler could have optimized the following a bit better:
```c
int main(void) {
    printf("Hello world %s\n", "");
}
```
Godbolt shows that using `ARM GCC 14.1.0` with options `-O3 -Os -Wall`, results in:
```asm
.LC0:
        .ascii  "\000"
.LC1:
        .ascii  "Hello world %s\012\000"
main:
        push    {r3, lr}
        ldr     r1, .L2
        ldr     r0, .L2+4
        bl      printf
        movs    r0, #0
        pop     {r3, pc}
.L2:
        .word   .LC0
        .word   .LC1
```.
That is, the empty string is not optimized away

alwa-nordic commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -3216,6 +3216,152 @@ struct bt_hci_evt_le_biginfo_adv_report {
 #define BT_HCI_ERR_TOO_LATE                     0x46
 #define BT_HCI_ERR_TOO_EARLY                    0x47
 
+static inline const char *bt_hci_err_to_str(uint8_t err)
```
Is the toolchain able to de-duplicate this function? If it is instantiated once for every compilation unit it's used in and doesn't optimize, then we should move the function definition to a implementation file.

rugeGerritsen replied:
Started out by creating https://github.com/zephyrproject-rtos/zephyr/pull/73826

LingaoM commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -3216,6 +3216,152 @@ struct bt_hci_evt_le_biginfo_adv_report {
+static inline const char *bt_hci_err_to_str(uint8_t err)
+{
+	switch (err) {
+	case BT_HCI_ERR_SUCCESS:
+		return "BT_HCI_ERR_SUCCESS(0x00)";
```
```suggestion
	case BT_HCI_ERR_SUCCESS:
		return "BT_HCI_ERR_SUCCESS(" STRINGIFY(BT_HCI_ERR_SUCCESS) ")";
```

rugeGerritsen committed:
Bluetooth: Conditionally print out HCI error codes as strings

When developing Bluetooth applications, you typically run into
some errors. If you are an experienced Bluetooth developer,
you would typically have an HCI error lookup table in your memory.
Others might not.

This commit utilizes defines CONFIG_BT_DEBUG_HCI_ERR_TO_STR
and utilizes bt_hci_err_to_str() to print out HCI error strings
when enabled to improve the user experience.

Several alternatives where considered. This approach was chosen
as it had the best balance between readability, code size, and
implementation complexity.

The alternatives are listed below as a reference.

1. Macro defined format specifier:

```c
  #define HCI_ERR_FMT "%s"
  #define BT_HCI_ERR_TO_STR(err) (err)
  #define HCI_ERR_FMT "%d"
  #define BT_HCI_ERR_TO_STR(err) bt_hci_err_to_str((err))

LOG_INF("The event contained " HCI_ERR_FMT " as status",
	BT_HCI_ERR_TO_STR(err));
```
Advantage: Space efficient: Code size does not increase
Disadvantage: Code becomes hard to read

2. Format specifier to always include both integer and string:

```c
static inline const char bt_hci_err_to_str(err)
{
	return "";
}

LOG_INF("The event contained %s(0x%02x) as status",
	bt_hci_err_to_str(err), err);
```

Advantage: Simple to use, implement, and read,
Disadvantage: Increases code size when CONFIG_BT_DEBUG_HCI_ERR_TO_STR
is disabled. The compiler seems unable to optimize away the unused
format specifier. Note: The size increase is only present when
logging is enabled.

3. Always print as string, allocate a stack variable when printing:

```c
const char *bt_hci_err_to_str(char *dst, size_t dst_size, uint8_t err)
{
  snprintf(dst, dst_size, 0x%02x, err);
  return dst;
}

LOG_INF("The event contained %s as status", BT_HCI_ERR_TO_STR(err));
```

Advantage: Very easy to read.
Disadvantage: Printing error codes becomes slow as it involves calling
snprint.

4. Implement a custom printf specifier, for example E.

   This requires a global CONFIG_ERR_AS_STR as I assume we cannot have
   one specifier for each type of error code.
   Also, I assume we cannot start adding specifiers for each subsystem.

```c
  #define BT_HCI_ERR_TO_STR(err) (err)
  #define BT_HCI_ERR_TO_STR(err) bt_hci_err_to_str((err))

LOG_INF("The event contained %E as status", BT_HCI_ERR_TO_STR(err));
```

Advantage: Both efficient code and readable code.
Disadvantage: This requires a global CONFIG_ERR_AS_STR as I assume
we cannot have one specifier for each type of error code.
Also, I assume we cannot start adding specifiers for each subsystem.
That is, this approach is hard to implement correctly in a scalable
way.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/ecc.c:
```diff
@@ -207,7 +207,7 @@ void bt_hci_evt_le_pkey_complete(struct net_buf *buf)
 	struct bt_pub_key_cb *cb;
 
-	LOG_DBG("status: 0x%02x", evt->status);
+	LOG_DBG("status: %s(0x%02x)", bt_hci_err_to_str(evt->status), evt->status);
```
Did we ever consider making the functions like `bt_hci_err_to_str` similar to `bt_addr_le_to_str`? i.e. it returns a value consisting of both a static string, and a variable value so that the string returned by `bt_hci_err_to_str` is always something like `BT_HCI_ERR_AUTH_FAIL(0x05)` so we don't need to add this type of formatting everywhere?

Thalley replied:
And of course in the case of `CONFIG_BT_DEBUG_HCI_ERR_TO_STR=n` it would just return `0x05`

rugeGerritsen replied:
> Did we ever consider making the functions like bt_hci_err_to_str similar to bt_addr_le_to_str? 
No I didn't. This sounds like a much better approach. Thanks for the suggestion!

rugeGerritsen replied:
Just to check if I've understood your proposal, here is how it would look:

```c
LOG_DBG("Status is: %s", bt_hci_err_to_str(status));
...

#ifdef CONFIG_BT_DEBUG_HCI_ERR_TO_STR
const char *bt_hci_err_to_str(uint8_t err)
{
  /* Existing implementation */
}
#else
const char *bt_hci_err_to_str(uint8_t err)
{
  static char err_as_hex[5];
  sprintf(err_as_hex, "0x%02x", err);
  return err_as_hex;
}
#endif
```

This will add some overhead, but probably not too much.

From what I've understood, the logging subsystem handles this nicely.

rugeGerritsen replied:
It would all have been simple if we converted a constant to a string instead of a variable. I agree that this is worse that we have today. I'm not sure how to make the tradeoff here

Thalley replied:
Indeed. It would have been trivial to do with the function was along the lines of 
```c
const char *bt_hci_err_to_str(uint8_t hci_err)
{
	#define HCI_ERR(err) [err] = #err

	const char * const mapping_table[] = {
		[BT_HCI_ERR_SUCCESS] = 
#if CONFIG_BT_DEBUG_HCI_ERR_TO_STR
			HCI_ERR(BT_HCI_ERR_SUCCESS)
#endif
			"(0x00)",
		[BT_HCI_ERR_UNKNOWN_CMD] = 
#if CONFIG_BT_DEBUG_HCI_ERR_TO_STR
			"BT_HCI_ERR_UNKNOWN_CMD"
#endif
			"(0x01)",
		...
}
```

So that `bt_hci_err_to_str` would either return the value or the full string

jori-nordic commented on subsys/bluetooth/common/bt_str.c:
```diff
@@ -141,8 +144,18 @@ const char *bt_hci_err_to_str(uint8_t hci_err)
 		return mapping_table[hci_err];
 	} else {
-		return "(unknown)";
+		static char err_as_hex[17];
```
where does the 17 come from?

alwa-nordic commented on subsys/bluetooth/common/bt_str.c:
```diff
@@ -141,8 +144,18 @@ const char *bt_hci_err_to_str(uint8_t hci_err)
+#else
+const char *bt_hci_err_to_str(uint8_t err)
+{
+	static char err_as_hex[5];
```
This has to at least be a thread local variable to be thread safe. We are committing to thread-safety as much as we can for new API.

Isn't this a bad pattern anyway? Returning a shared resource without a lock. Especially in public API, this can be very confusing for users. How do we document this API?

I think it's preferable to take the destination buffer as an argument, sprintf-like.

alwa-nordic replied:
To consider: We can make the function not allocate and instead take the destination buffer as input, a like `snprintf`. Unlike `snprintf`, we return the same pointer. This allows using the function directly in log statements. Finally, we add a macro that allocates an automatic variable for ease of use.

```c
#define BT_HCI_ERR_TO_STR(err) bt_hci_err_to_str(&(char[16]){}, 16, (err))
const char *bt_hci_err_to_str(char *dst, size_t dst_size, uint8_t err);
```

Thalley replied:
That is a good solution, although it requires the allocation of that string before each log statement. Given that the resulting string is part of a limited set of possibilities, have a constant lookup table would still be better IMO, as we don't really need to do anything dynamically (unless we want "unknown(0xYY)").

Thalley replied:
> I thought that it might be a good idea to add an error code when it is unknown to make it clear what is going wrong. Maybe we can simply log this separately?
> 
> ```c
> const char *bt_hci_err_to_str(uint8_t hci_err)
> {
>   	const char * const mapping_table[] = {ENTRIES};
> 	if (hci_err < ARRAY_SIZE(mapping_table) && mapping_table[hci_err]) {
> 		return mapping_table[hci_err];
> 	} else {
> 		LOG_ERR("Unknown HCI error code translation %d", hci_err);
> 		return "(unknown)";
> 	}
> }
> ```
> 
> Should we keep the `bt_hci_err_to_str()` variant that converts to a hex representation?

If we want to avoid that original suggestion with `%s (0x%02X)` then I think this would be the best solution, assuming that `mapping_table` here contains the entirety of what we want to get. 

Otherwise the original suggestion was never bad, just that it would be nice to be able to always just do `%s`

alwa-nordic commented:
It would be nice to have [custom printf specifier](https://www.gnu.org/software/libc/manual/html_node/Customizing-Printf.html) support in Zephyr.

rugeGerritsen commented:
> It would be nice to have [custom printf specifier](https://www.gnu.org/software/libc/manual/html_node/Customizing-Printf.html) support in Zephyr.

I thought about this too. But that probably requires more changes as it requires updates to all packages that use format specifiers

rugeGerritsen commented:
It seems like we end up with the following options:
1. Add a define for the format specifier (either `%s` or `%d`). This creates code that looks a bit ugly: 
```c
#ifdef CONFIG_ERR_AS_STR
  #define ERR_FMT "%s"
#else
  #define ERR_FMT "%d"
#endif

LOG_INF("The event contained " ERR_FMT " as status", bt_hci_err_to_str(err));
```

2. Always print both, but where the string is sometimes empty: 
```c
LOG_INF("The event contained %s(0x%02x) as status", bt_hci_err_to_str(err), err);
```

3. Allocate a stack variable to be used with `snprintf` each time we want to print a error code as a hex:
```c
#define BT_HCI_ERR_TO_STR(err) bt_hci_err_to_str(&(char[16]){}, 16, (err))
const char *bt_hci_err_to_str(char *dst, size_t dst_size, uint8_t err);

LOG_INF("The event contained %s as status", BT_HCI_ERR_TO_STR(err));
```

4. Implement a custom printf specifier, for example `E`. This requires a global `CONFIG_ERR_AS_STR` as I assume we cannot have one specifier for each type of error code. Also, I assume we cannot start adding specifiers for each subsystem.
```c
LOG_INF("The event contained %E as status", bt_hci_err_to_str(err));
```

I believe option 2 or 3 is the way to go. Option 2 is the cheaper of those two

Thalley commented:
> I believe option 2 or 3 is the way to go. Option 2 is the cheaper of those two

I agree that of those, option 2 and 3 are best. 

My suggestion from https://github.com/zephyrproject-rtos/zephyr/pull/73795#discussion_r1648972422 would have the size benefit of option 2 and the ease of use of option 3, but at the cost of not being able to log the unknown HCI error (but also with the disadvantage of looking terrible in the source code :upside_down_face: )

rugeGerritsen commented:
> > I believe option 2 or 3 is the way to go. Option 2 is the cheaper of those two
> 
> I agree that of those, option 2 and 3 are best.
> 
> My suggestion from [#73795 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/73795#discussion_r1648972422) would have the size benefit of option 2 and the ease of use of option 3, but at the cost of not being able to log the unknown HCI error (but also with the disadvantage of looking terrible in the source code 🙃 )

Also, when string representations are not compiled in, all error codes are unknown, which I believe is a bigger drawback

Thalley commented:
> Also, when string representations are not compiled in, all error codes are unknown, which I believe is a bigger drawback

Can you elaborate on that? 

In my suggest if `CONFIG_BT_DEBUG_HCI_ERR_TO_STR=n` it would still return `"(0x00)"` for `BT_HCI_ERR_SUCCESS`, "(0x01)" for `BT_HCI_ERR_UNKNOWN_CMD`, etc.

alwa-nordic commented:
>     4. Implement a custom printf specifier, for example `E`. This requires a global `CONFIG_ERR_AS_STR` as I assume we cannot have one specifier for each type of error code. Also, I assume we cannot start adding specifiers for each subsystem.

I suspect it's possible to hack it so we only use one letter code for multiple error types. I'm thinking something like the code below can be interpreted by a `Z` (for Zephyr) specifier in a `print_widget_arginfo`.

```c
printf("Hello %hcierrZ world", err);
```

See https://www.gnu.org/software/libc/manual/html_node/Printf-Extension-Example.html.

alwa-nordic commented:
I am for option 2.

alwa-nordic commented:
I would like to discuss that we should have a more complete tracing component. It should be able to take arbitrary data types/fields and be able to optionally output to the log with a format custom for that field.

Thalley commented:
> I would like to discuss that we should have a more complete tracing component. It should be able to take arbitrary data types/fields and be able to optionally output to the log with a format custom for that field.

Sounds like you are thinking something similar to [Display](https://doc.rust-lang.org/std/fmt/trait.Display.html) in Rust or `__str__` in Python? 

It sounds good, but also sounds ambitious (and will likely get some pushback from the C purists)

Thalley commented:
> I am for option 2.

Of 2 and 3, I would also prefer 2 if 3 cannot be done without a local variable

rugeGerritsen commented:
> > Also, when string representations are not compiled in, all error codes are unknown, which I believe is a bigger drawback
> 
> Can you elaborate on that?
> 
> In my suggestion if `CONFIG_BT_DEBUG_HCI_ERR_TO_STR=n` it would still return `"(0x00)"` for `BT_HCI_ERR_SUCCESS`, "(0x01)" for `BT_HCI_ERR_UNKNOWN_CMD`, etc.

When `CONFIG_BT_DEBUG_HCI_ERR_TO_STR=n`, we still need the format string to use `%s`. This means we either need a mapping table, or generate the string on the fly:
```c
const char *bt_hci_err_to_str(uint8_t err)
{
  const char *mapping_table[] = {"0x00", 0x01, 0x02, ...");
  return mapping_table[err];
}

LOG_INF("status is %s", bt_hci_err_to_str(err));
```
The only case where we would not need a mapping table, is when the error code is a compile time constant. That is usually not the case:

```c
#define bt_hci_err_to_str(err) #err

LOG_INF("status is %s", bt_hci_err_to_str(BT_HCI_ERR_SUCCESS));
```

We could of course argue that we should add a constant mapping table containing all known error codes as hex, but that is also quite space consuming

rugeGerritsen commented:
Pushed the implementation of alternative 2. The commit message elaborates why this was chosen.

Thalley commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -32,7 +32,11 @@ extern "C" {
  *
  * @return The string representation of the HCI error code.
  */
+#ifdef CONFIG_BT_DEBUG_HCI_ERR_TO_STR
 const char *bt_hci_err_to_str(uint8_t hci_err);
+#else
+static inline const char *bt_hci_err_to_str(uint8_t hci_err) { return ""; }
+#endif
```
If the implementation of `bt_hci_err_to_str` just returned `""` instead of this "hack", does that increase memory usage significantly?

rugeGerritsen replied:
It increases memory consumption a little bit. For the sample peripheral_hr on 52840dk with the default configuration (no debug logs):

With "optimization":
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      158376 B         1 MB     15.10%
             RAM:       25748 B       256 KB      9.82%
        IDT_LIST:          0 GB        32 KB      0.00%
```
Without (+16 bytes)
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      158392 B         1 MB     15.11%
             RAM:       25748 B       256 KB      9.82%
        IDT_LIST:          0 GB        32 KB      0.00%
```

Before this commit (-156)
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      158236 B         1 MB     15.09%
             RAM:       25748 B       256 KB      9.82%
        IDT_LIST:          0 GB        32 KB      0.00%
```

Thalley replied:
Is that with `-Os`? 

I'm OK to spend 16 bytes of flash to just have a single definition of `bt_hci_err_to_str`

rugeGerritsen replied:
Yes, this is with default compiler options

Thalley commented on subsys/bluetooth/Kconfig.debug:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config BT_DEBUG_HCI_ERR_TO_STR
+	bool "Print HCI error codes as strings"
+	help
+	  This configuration enables printing of HCI error
+	  codes represented as strings.
+	  See bt_hci_err_to_str() for more details.
```
Is `DEBUG` really useful to denote? 

Wouldn't it work equally well if it was just `BT_HCI_ERR_TO_STR`? 

The function can be used outside of debug purposes

jori-nordic commented on subsys/bluetooth/Kconfig.debug:
```diff
@@ -0,0 +1,11 @@
+
+config BT_DEBUG_HCI_ERR_TO_STR
+	bool "Print HCI error codes as strings"
+	help
```
can you add an `select EXPERIMENTAL`?
Then we can just merge your proposal and iterate after some real-world usage. I don't think this functionality is going to break much.

rugeGerritsen replied:
I believe we should merge https://github.com/zephyrproject-rtos/zephyr/pull/75489 as part of the 3.7.0 release to avoid changing the API

Thalley commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -32,7 +32,11 @@ extern "C" {
  *
  * @return The string representation of the HCI error code.
  */
+#ifdef CONFIG_BT_HCI_ERR_TO_STR
 const char *bt_hci_err_to_str(uint8_t hci_err);
+#else
+static inline const char *bt_hci_err_to_str(uint8_t hci_err) { return ""; }
+#endif
```
Since it's now guarded, we should probably add a note in the documentation for it:

```
 * If @kconfig{CONFIG_BT_HCI_ERR_TO_STR} is not enabled, this just returns the empty string
```

Or something like that

Thalley commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -32,7 +32,11 @@ extern "C" {
  *
  * @return The string representation of the HCI error code.
  */
+#ifdef CONFIG_BT_HCI_ERR_TO_STR
```
```suggestion
#if defined(CONFIG_BT_HCI_ERR_TO_STR)
```

Suggest to be consistent with the .c file

fabiobaltieri commented:
merge conflict, can you rebase please?

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -432,7 +432,8 @@ int bt_hci_cmd_send_sync(uint16_t opcode, struct net_buf *buf,
 	status = cmd(buf)->status;
 	if (status) {
-		LOG_WRN("opcode 0x%04x status 0x%02x", opcode, status);
+		LOG_WRN("opcode 0x%04x status %s(0x%02x)", opcode,
+			bt_hci_err_to_str(status), status);
```
In https://github.com/zephyrproject-rtos/zephyr/pull/75442 the formatting is `%u %s`. Should we use the same formatting here?

rugeGerritsen replied:
Yes, I agree. We should should change it to that format
